[Rework] Blue Magic

ShiyoKozuki opened the pull request:
-Adds the ability to add attack bonus to spells
- Adds all breath spells proper scaling formulas based on HP
- Adds resist checks to BLU additional effects
- Adds correlation bonus to spells
- Adds acc bonus to all phys spells
- Adds convergence
- Adds monster correlation
- Adds proper message for when a physical spell fully misses
- Adds proper half/ full/ no resist status effect example

- Adds crit varies with TP
- Adds attack varies with TP
- Adds duration varies with TP
- Adds accuracy varies with TP

This is just the framework to work with, I did not fix every single spell but I gave the tools and formulas to adjust the rest easily.

Sources:
https://www.bg-wiki.com/ffxi/Blue_Mage
https://www.bluegartr.com/threads/37619-Blue-Mage-Best-thread-ever
http://wiki.ffo.jp/html/1726.html
http://www.neko-sentai.com/images/BlumageNFO.htm

<!-- remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm: -->
**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] that I agree to LandSandBoat's [Limited Contributor License Agreement](https://github.com/LandSandBoat/server/blob/base/.github/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [x] that I have **read the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)**
- [x] that I've _**tested my code and things my code changed**_ since the last commit in the PR, and will test after any later commits

ShiyoKozuki committed:
[Rework] Blue Magic

-Adds the ability to add attack bonus to spells
- Adds all breath spells proper scaling formulas based on HP
- Adds resist checks to BLU additional effects
- Adds correlation bonus to spells
- Adds acc bonus to all phys spells
- Adds convergencebonus
- Adds monster correlation
- Adds proper message for when a physical spell fully misses
- Adds proper half/ full/ no resist status effect example

- Adds crit varies with TP
- Adds attack varies with TP
- Adds duration varies with TP
- Adds accuracy varies with TP

This is just the framework to work with, I did not fix every single spell but I gave the tools and formulas to adjust the rest easily.

claywar commented on scripts/globals/bluemagic.lua:
```diff
@@ -9,186 +9,28 @@ TPMOD_ACC = 3
 TPMOD_ATTACK = 4
 TPMOD_DURATION = 5
 
+-- The SC the spell makes
```
These should not exist in a global space unattached to any object.  They were removed initially because they also conflict with weaponskill definitions.  If really needed as a global, they should be tabled and attached to the xi global.

ShiyoKozuki replied:
Do you want all of them gone or just TPMOD ones? I don't think any of them do anything, at least TPMOD ones seem to do nothing I think?

Xaver-DaRed replied:
At a future date, we may want blue spells to have their properties tabled. For now, though, it isnt worse than what we have now.

claywar commented on scripts/globals/bluemagic.lua:
```diff
@@ -263,37 +105,76 @@ function BluePhysicalSpell(caster, target, spell, params)
+    ----------------------------------------------
+    -- Get the possible pDIF range and hit rate --
+    ----------------------------------------------
+    if (params.attkbonus == nil) then
```
Just some minor styling, but parens are not necessary here, or any case outside of an and/or grouping.

ShiyoKozuki replied:
When I was testing it returned as "nil" and errors unless I put it in every lua. This allows you to not have to add it into every lua with a value since attkbonus isn't used in every single spell.

Or am I wrong?

claywar replied:
Parenthesis should have no impact on the returned value.  That's what I'm referring to.

claywar commented on scripts/globals/bluemagic.lua:
```diff
@@ -369,22 +239,27 @@ function BlueMagicalSpell(caster, target, spell, params, statMod)
     magicAttack = math.floor(magicAttack * applyResistance(caster, target, spell, rparams))
 
-    local dmg = math.floor(addBonuses(caster, spell, target, magicAttack))
+    dmg = math.floor(addBonuses(caster, spell, target, magicAttack))
```
Why is dmg being set as global here? Local should still apply.

TeoTwawki commented on scripts/globals/spells/bluemagic/disseverment.lua:
```diff
@@ -1,4 +1,4 @@
------------------------------------
+-----------------------------------------
```
you should avoid any extra changes like this for the time being. just adds to an already large diff to read.

claywar commented on scripts/globals/bluemagic.lua:
```diff
@@ -400,36 +275,233 @@ function BlueFinalAdjustments(caster, target, spell, dmg, params)
 end
+
+-- obtains alpha, used for working out WSC
+function BlueGetAlpha(level)
```
This feels like it could be normalized into a mathematical function.

claywar reviewed (changes requested):
Great first effort, but there's a lot of cleanup that can happen as well.  Please use the style guide in the wiki as a ref, but in all, I look forward to seeing how this progresses!

Requesting Xaver to review as well, since he has been doing a ton of magic refactoring.  Also, there's a lot of items in CI that need to be addressed.

ShiyoKozuki committed:
[BLU Rework] Fixes lua errors

ShiyoKozuki committed:
Fixes

ShiyoKozuki committed:
Style changes

ShiyoKozuki committed:
Convergence fix

almuth150 commented:
Did you come across anything about how Subtle Blow affects TP feed for BLU spells? According to [the trait page on BG Wiki](https://www.bg-wiki.com/ffxi/Subtle_Blow), it says that TP feed is reduced by the trait in general, so I'm pretty sure it should also affect BLU spells. AFAIK this is not implemented at all, however.

Xaver-DaRed commented:
I dont see anything weird or overly problematic.
Its a good first effort, undoubtedly.
Blue magic still needs a rewrite, but for now, adding missing functionality and documentation is a good thing.

Do correct the style guide issues, though.
Remember, no unneeded "( )" and 4 spaces. no tabs. Text properly aligned.

Xaver-DaRed reviewed (changes requested):
Correct style, mainly.

Also, dont be afraid to comment the code. The more comments, the easier to know whats going on, visualy.

Snugglesffxi commented:
@ShiyoKozuki you coming back to this? (no rush) we have some very excited blue mages esp since their beloved broken ass mindblast has been "nerfed by snuggles" ;-;

ShiyoKozuki committed:
Ranged weapon skill acc level penalty fix

ShiyoKozuki committed:
Delete weaponskills.lua

ShiyoKozuki committed:
Revert "Ranged weapon skill acc level penalty fix"

This reverts commit 201683096b5e1308a95b25ace36c257f84cdd0fe.

ShiyoKozuki committed:
Merge branch 'base' of https://github.com/ShiyoKozuki/server into base

ShiyoKozuki committed:
Create weaponskills.lua

ShiyoKozuki committed:
[BLU Rework] Add subtle blow to blue magic

ShiyoKozuki committed:
[Blu Magic Revamp] Fixes a lot of stuff that doesn't work and changed formulas

ShiyoKozuki committed:
Merge branch 'base' into base

ShiyoKozuki committed:
[Blue Magic Revamp] Fixes

ShiyoKozuki committed:
Merge branch 'base' of https://github.com/ShiyoKozuki/server into base

ShiyoKozuki committed:
[BlU Rework] Add shadow ignore/wipe to spells that should have it and fix blue phys attack formula

ShiyoKozuki committed:
[Blu Rework] Add source

ShiyoKozuki committed:
[BLU Rework] Regurg Enmity

ShiyoKozuki committed:
Corossive Ooze and self destruct fixes

Corossive ooze is static enmity like regurg. Self destruct shouldn't bypass resistance checks.

Skerxan commented:
Bluemage needs all the love it can get ; ;

Xaver-DaRed commented:
I'm going to close this due to Author's innactivity.

Shiyo, if you ever decide to come back and finish this PR, feel free to re-open it.
